File: paleo_workbench/mapping_workspace/constraints_sync.py
"""Constraint geometry sync-back (goal §11, audit P0-3).

``create_constraint`` (ui/workstation/stage_actions.py) registers an editable
``UserVectorLayer`` plus a ``ConstraintLine`` whose ``coordinates`` are EMPTY —
geometry is digitized into the vector layer, and the engine-side line used by
interpolation could silently diverge.  This module harvests the digitized
features back into the linked ``ConstraintLine`` entries and stamps a content
fingerprint so downstream freshness evaluation has a comparison baseline
(``constraints:current``).

Matching rule (conservative, machine-readable signals only):

1. primary — ``ConstraintLine.properties["layer_id"] == layer_id`` (the stamp
   ``create_constraint`` writes);
2. fallback — same ``constraint_kind`` AND the line name equals the layer
   name AND the line still has empty coordinates (legacy layers created
   before the ``layer_id`` stamp existed).

Harvest semantics:

* line-kind layers — one coordinate sequence per feature (MultiLineString
  parts are joined, dropping duplicated joint vertices);
* polygon kinds (mask / exclusion area) — the exterior ring of each feature,
  **closed** (first vertex appended when the digitizer left it open; the
  interpolation ring consumer requires closed rings);
* a layer with multiple features syncs one ``ConstraintLine`` per feature
  (the first matched line keeps its identity; extras are appended with the
  same ``layer_id`` stamp); stale extra lines from a previous sync are
  replaced, never accumulated;
* every synced line's ``properties`` gains ``{"content_fingerprint":
  sha256(coordinates)}``.

Wiring: called from ``StageActionDispatcher.stage_save`` (after
``flush_edit_sessions`` has committed the vector layers into
``ProjectDocument.user_vector_layers``).  It is deliberately a pure document
transform — no Qt — so readiness/overlay paths can also call it directly.
"""
from __future__ import annotations

import hashlib
import json
import math
from typing import Any, Iterable, Mapping

from paleo_workbench.mapping_workspace.layer_roles import (
    constraint_kind_from_value,
)

__all__ = [
    "constraint_content_fingerprint",
    "sync_constraint_geometry",
]


def _finite(value: Any) -> float | None:
    try:
        number = float(value)
    except (TypeError, ValueError):
        return None
    return number if math.isfinite(number) else None


def _vertex(point: Any) -> list[float] | None:
    if not isinstance(point, (list, tuple)) or len(point) < 2:
        return None
    x = _finite(point[0])
    y = _finite(point[1])
    if x is None or y is None:
        return None
    return [x, y]


def constraint_content_fingerprint(coordinates: Iterable[Iterable[float]]) -> str:
    """sha256 over a canonical encoding of the coordinate sequence.

    Same coordinates → same hash (floats are rounded to 9 decimals so trivial
    representation noise does not churn the fingerprint).
    """
    canonical = [
        [round(float(point[0]), 9), round(float(point[1]), 9)]
        for point in coordinates
        if _vertex(point) is not None
    ]
    payload = json.dumps(canonical, sort_keys=True, separators=(",", ":"))
    return hashlib.sha256(payload.encode("utf-8")).hexdigest()


def _line_kind_of(layer: Any) -> bool:
    kind = constraint_kind_from_value(str(getattr(layer, "template", "") or ""))
    if kind is not None:
        return kind.geometry_kind != "polygon"
    return str(getattr(layer, "geometry_kind", "line") or "line") != "polygon"


def _harvest_coordinates(layer: Any) -> tuple[list[list[list[float]]], int]:
    """Feature geometries → coordinate sequences (closed rings for polygons).

    Returns ``(sequences, holes_seen)`` where holes_seen counts interior
    rings skipped by polygon harvesting (the interpolation ring consumer
    uses exterior domains only).
    """
    is_line = _line_kind_of(layer)
    sequences: list[list[list[float]]] = []
    holes = 0
    for feature in getattr(layer, "features", None) or []:
        geometry = getattr(feature, "geometry", None)
        if not isinstance(geometry, Mapping):
            continue
        gtype = str(geometry.get("type") or "")
        coordinates = geometry.get("coordinates")
        if not coordinates:
            continue
        if is_line:
            parts: list[Any] = []
            if gtype == "LineString":
                parts = [coordinates]
            elif gtype == "MultiLineString":
                parts = list(coordinates)
            else:
                continue
            merged: list[list[float]] = []
            for part in parts:
                for point in part or []:
                    vertex = _vertex(point)
                    if vertex is None:
                        continue
                    if merged and merged[-1] == vertex:
                        continue  # drop duplicated joints between parts
                    merged.append(vertex)
            if len(merged) >= 2:
                sequences.append(merged)
        else:
            if gtype == "Polygon":
                rings = list(coordinates)
            elif gtype == "MultiPolygon":
                rings = [ring for poly in coordinates for ring in poly]
            else:
                continue
            for index, ring in enumerate(rings):
                points: list[list[float]] = []
                for point in ring or []:
                    vertex = _vertex(point)
                    if vertex is not None:
                        points.append(vertex)
                if len(points) < 3:
                    continue
                if index == 0:
                    # Ring closure for polygon kinds (mask / exclusion):
                    # the boundary-ring consumer requires first == last.
                    if points[0] != points[-1]:
                        points.append(list(points[0]))
                    sequences.append(points)
                else:
                    holes += 1
    return sequences, holes


def _linked_lines(document: Any, layer: Any, layer_id: str) -> tuple[list, str]:
    """Find the ConstraintLine entries linked to this vector layer."""
    kind_value = str(getattr(layer, "template", "") or "")
    layer_name = str(getattr(layer, "name", "") or "")
    by_layer_id: list = []
    by_kind: list = []
    for group in getattr(document, "constraint_layers", None) or []:
        for line in group.lines:
            properties = line.properties or {}
            if str(properties.get("layer_id") or "") == layer_id:
                by_layer_id.append(line)
            elif (
                kind_value
                and str(properties.get("constraint_kind") or "") == kind_value
                and str(line.name or "") == layer_name
                and not line.coordinates
            ):
                by_kind.append(line)
    if by_layer_id:
        return by_layer_id, "layer_id"
    if by_kind:
        return by_kind, "constraint_kind+name"
    return [], ""


def _target_group(document: Any, matched: list):
    for group in getattr(document, "constraint_layers", None) or []:
        if any(line in group.lines for line in matched):
            return group
    groups = getattr(document, "constraint_layers", None)
    if groups is None:
        return None
    if groups:
        return groups[0]
    from paleo_workbench.project.models import ConstraintLayers

    group = ConstraintLayers(name="约束层")
    document.constraint_layers.append(group)
    return group


def sync_constraint_geometry(document: Any, layer_id: str) -> dict:
    """Harvest a constraint ``UserVectorLayer``'s features into the linked
    ``ConstraintLine.coordinates`` (+ content fingerprint).

    Returns a report dict; ``ok`` is False with a ``reason`` when nothing
    was synced (unknown layer / not a constraint layer / empty layer).
    """
    layer_id = str(layer_id or "")
    layer = None
    for candidate in getattr(document, "user_vector_layers", None) or []:
        if str(getattr(candidate, "id", "") or "") == layer_id:
            layer = candidate
            break
    if layer is None:
        return {"ok": False, "reason": "layer_not_found", "layer_id": layer_id}
    kind = constraint_kind_from_value(str(getattr(layer, "template", "") or ""))
    if kind is None:
        return {
            "ok": False,
            "reason": "not_a_constraint_layer",
            "layer_id": layer_id,
        }
    sequences, holes = _harvest_coordinates(layer)
    if not sequences:
        # Never wipe previously synced geometry based on an empty transient
        # state (digitizing may still be in progress).
        return {
            "ok": False,
            "reason": "layer_empty",
            "layer_id": layer_id,
            "constraint_kind": kind.value,
        }

    matched, matched_by = _linked_lines(document, layer, layer_id)
    group = _target_group(document, matched)
    if group is None:
        return {"ok": False, "reason": "no_constraint_group", "layer_id": layer_id}

    template_line = matched[0] if matched else None
    # Replace semantics: stale extra lines from a previous sync are dropped
    # (never accumulated), then one line per harvested sequence is appended.
    stale = set(id(line) for line in matched)
    group.lines = [line for line in group.lines if id(line) not in stale]

    from paleo_workbench.project.models import ConstraintLine

    synced: list[ConstraintLine] = []
    for index, coordinates in enumerate(sequences):
        fingerprint = constraint_content_fingerprint(coordinates)
        if index == 0 and template_line is not None:
            line = template_line
            line.coordinates = coordinates
        else:
            line = ConstraintLine(
                name=str(getattr(layer, "name", "") or kind.label),
                role=template_line.role if template_line is not None
                else kind.interpolation_role,
                target_horizon=(
                    template_line.target_horizon if template_line is not None
                    else str(getattr(group, "target_horizon", "") or "")
                ),
                properties={
                    "layer_id": layer_id,
                    "constraint_kind": kind.value,
                    "feature_index": index,
                },
            )
            line.coordinates = coordinates
        group.lines.append(line)
        properties = dict(line.properties or {})
        properties["layer_id"] = layer_id
        properties["constraint_kind"] = kind.value
        properties["content_fingerprint"] = fingerprint
        line.properties = properties
        synced.append(line)

    aggregate = constraint_content_fingerprint(
        [point for coordinates in sequences for point in coordinates])
    return {
        "ok": True,
        "layer_id": layer_id,
        "constraint_kind": kind.value,
        "matched_by": matched_by or "created_new_line",
        "lines_synced": len(synced),
        "features_harvested": len(sequences),
        "interior_rings_skipped": holes,
        "coordinates_total": sum(len(seq) for seq in sequences),
        "content_fingerprint": aggregate,
    }

File: libs/ui_composite/src/constraints_sync.cpp
#include "pwb/ui_composite/constraints_sync.hpp"

#include "pwb/domain/sha256.hpp"
#include "pwb/ui_composite/roles.hpp"

#include <cmath>
#include <utility>
#include <vector>

namespace pwb::ui_composite {
namespace {

std::optional<double> finite(const Json& value) {
    double number;
    if (value.is_number()) {
        number = value.get<double>();
    } else if (value.is_string()) {
        try {
            number = std::stod(value.get<std::string>());
        } catch (const std::exception&) {
            return std::nullopt;
        }
    } else {
        return std::nullopt;
    }
    return std::isfinite(number) ? std::optional<double>{number}
                                 : std::nullopt;
}

// _vertex parity — [x, y] or invalid.
Json vertex_of(const Json& point) {
    if (!point.is_array() || point.size() < 2) return Json();
    const auto x = finite(point[0]);
    const auto y = finite(point[1]);
    if (!x || !y) return Json();
    return Json::array({*x, *y});
}

Json* field(Json& object, const char* key) {
    if (!object.is_object()) return nullptr;
    const auto it = object.find(key);
    return it != object.end() ? &*it : nullptr;
}

const Json* field(const Json& object, const char* key) {
    if (!object.is_object()) return nullptr;
    const auto it = object.find(key);
    return it != object.end() ? &*it : nullptr;
}

std::string str_of(const Json& object, const char* key) {
    const Json* v = field(object, key);
    if (v == nullptr || v->is_null()) return "";
    if (v->is_string()) return v->get<std::string>();
    if (v->is_boolean()) return v->get<bool>() ? "True" : "False";
    if (v->is_number_integer()) return std::to_string(v->get<long long>());
    if (v->is_number_unsigned())
        return std::to_string(v->get<unsigned long long>());
    if (v->is_number()) return std::to_string(v->get<double>());
    return "";
}

// Round to 9 decimals (Python round(float(x), 9) — round-half-even on
// the decimal representation; std::round is half-away but the 1e-9 scale
// only ever differs on exact .5 ties of the 9th decimal, which the
// Python path hits identically rarely — kept simple per the docstring).
double round9(double value) {
    return std::round(value * 1e9) / 1e9;
}

bool line_kind_of(const Json& layer) {
    const std::string kind_value = str_of(layer, "template");
    const auto kind = constraint_kind_from_value(kind_value);
    if (kind.has_value()) {
        return constraint_kind_geometry_kind(*kind) != "polygon";
    }
    return str_of(layer, "geometry_kind") != "polygon";
}

// _harvest_coordinates parity — feature geometries → sequences +
// interior-rings-skipped count.
std::pair<std::vector<Json>, int> harvest_coordinates(const Json& layer) {
    const bool is_line = line_kind_of(layer);
    std::vector<Json> sequences;
    int holes = 0;
    const Json* features = field(layer, "features");
    if (features == nullptr || !features->is_array()) {
        return {sequences, holes};
    }
    for (const Json& feature : *features) {
        const Json* geometry = field(feature, "geometry");
        if (geometry == nullptr || !geometry->is_object()) continue;
        const std::string gtype = str_of(*geometry, "type");
        const Json* coordinates = field(*geometry, "coordinates");
        if (coordinates == nullptr || coordinates->empty()
            || !coordinates->is_array())
            continue;
        if (is_line) {
            std::vector<const Json*> parts;
            if (gtype == "LineString") {
                parts.push_back(coordinates);
            } else if (gtype == "MultiLineString") {
                for (const Json& part : *coordinates) parts.push_back(&part);
            } else {
                continue;
            }
            Json merged = Json::array();
            for (const Json* part : parts) {
                if (!part->is_array()) continue;
                for (const Json& point : *part) {
                    const Json vertex = vertex_of(point);
                    if (vertex.is_null()) continue;
                    if (!merged.empty() && merged.back() == vertex) {
                        continue;  // drop duplicated joints between parts
                    }
                    merged.push_back(vertex);
                }
            }
            if (merged.size() >= 2) sequences.push_back(std::move(merged));
        } else {
            std::vector<const Json*> polys;
            if (gtype == "Polygon") {
                polys.push_back(coordinates);
            } else if (gtype == "MultiPolygon") {
                for (const Json& poly : *coordinates) polys.push_back(&poly);
            } else {
                continue;
            }
            for (const Json* poly : polys) {
                if (!poly->is_array() || poly->empty()) continue;
                int index = 0;
                for (const Json& ring : *poly) {
                    Json points = Json::array();
                    if (ring.is_array()) {
                        for (const Json& point : ring) {
                            const Json vertex = vertex_of(point);
                            if (!vertex.is_null()) points.push_back(vertex);
                        }
                    }
                    ++index;
                    if (points.size() < 3) continue;
                    if (index == 1) {
                        // Ring closure for polygon kinds (mask /
                        // exclusion): the boundary-ring consumer requires
                        // first == last.
                        if (points.front() != points.back()) {
                            points.push_back(points.front());
                        }
                        sequences.push_back(std::move(points));
                    } else {
                        ++holes;
                    }
                }
            }
        }
    }
    return {sequences, holes};
}

// _linked_lines parity — (matched line Json*, matched_by). Lines are
// located inside document["constraint_layers"][*]["lines"].
std::pair<std::vector<Json*>, std::string> linked_lines(
    Json& document, const Json& layer, const std::string& layer_id) {
    const std::string kind_value = str_of(layer, "template");
    const std::string layer_name = str_of(layer, "name");
    std::vector<Json*> by_layer_id;
    std::vector<Json*> by_kind;
    Json* groups = field(document, "constraint_layers");
    if (groups != nullptr && groups->is_array()) {
        for (Json& group : *groups) {
            Json* lines = field(group, "lines");
            if (lines == nullptr || !lines->is_array()) continue;
            for (Json& line : *lines) {
                const Json* props = field(line, "properties");
                const Json empty = Json::object();
                const Json& p = props != nullptr ? *props : empty;
                if (str_of(p, "layer_id") == layer_id) {
                    by_layer_id.push_back(&line);
                } else if (!kind_value.empty()
                           && str_of(p, "constraint_kind") == kind_value
                           && str_of(line, "name") == layer_name) {
                    const Json* coords = field(line, "coordinates");
                    if (coords == nullptr || coords->empty()) {
                        by_kind.push_back(&line);
                    }
                }
            }
        }
    }
    if (!by_layer_id.empty()) return {by_layer_id, "layer_id"};
    if (!by_kind.empty()) return {by_kind, "constraint_kind+name"};
    return {{}, ""};
}

// _target_group parity — the group containing a matched line; else the
// first group; else a created "约束层" group appended to the document.
Json* target_group(Json& document, const std::vector<Json*>& matched) {
    Json* groups = field(document, "constraint_layers");
    if (groups != nullptr && groups->is_array()) {
        if (!matched.empty()) {
            for (Json& group : *groups) {
                Json* lines = field(group, "lines");
                if (lines == nullptr || !lines->is_array()) continue;
                for (Json& line : *lines) {
                    for (Json* m : matched) {
                        if (&line == m) return &group;
                    }
                }
            }
        }
        if (!groups->empty()) return &groups->front();
    }
    if (groups == nullptr || !groups->is_array()) {
        document["constraint_layers"] = Json::array();
        groups = &document["constraint_layers"];
    }
    Json group = Json::object();
    group["name"] = "约束层";
    group["lines"] = Json::array();
    groups->push_back(std::move(group));
    return &groups->back();
}

}  // namespace

std::string constraint_content_fingerprint(const Json& coordinates) {
    Json canonical = Json::array();
    if (coordinates.is_array()) {
        for (const Json& point : coordinates) {
            const Json vertex = vertex_of(point);
            if (vertex.is_null()) continue;
            canonical.push_back(
                Json::array({round9(vertex[0].get<double>()),
                             round9(vertex[1].get<double>())}));
        }
    }
    // json.dumps(canonical, sort_keys=True, separators=(",", ":")) —
    // arrays of numbers are key-free so ordering is already canonical.
    return domain::Sha256::of_bytes(canonical.dump());
}

Json sync_constraint_geometry(Json& document, const std::string& layer_id) {
    Json report = Json::object();
    report["ok"] = false;
    report["layer_id"] = layer_id;

    // Locate the vector layer inside user_vector_layers.
    Json* layer = nullptr;
    Json* layers = field(document, "user_vector_layers");
    if (layers != nullptr && layers->is_array()) {
        for (Json& candidate : *layers) {
            if (str_of(candidate, "id") == layer_id) {
                layer = &candidate;
                break;
            }
        }
    }
    if (layer == nullptr) {
        report["reason"] = "layer_not_found";
        return report;
    }
    const std::string kind_value = str_of(*layer, "template");
    const auto kind = constraint_kind_from_value(kind_value);
    if (!kind.has_value()) {
        report["reason"] = "not_a_constraint_layer";
        return report;
    }
    report["constraint_kind"] = *kind;

    auto [sequences, holes] = harvest_coordinates(*layer);
    if (sequences.empty()) {
        // Never wipe previously synced geometry based on an empty
        // transient state (digitizing may still be in progress).
        report["reason"] = "layer_empty";
        return report;
    }

    auto [matched, matched_by] = linked_lines(document, *layer, layer_id);
    Json* group = target_group(document, matched);
    if (group == nullptr) {
        report["reason"] = "no_constraint_group";
        return report;
    }

    Json* template_line = matched.empty() ? nullptr : matched.front();
    // Replace semantics: stale extra lines from a previous sync are
    // dropped (never accumulated), then one line per harvested sequence
    // is appended.
    Json* lines = field(*group, "lines");
    if (lines == nullptr || !lines->is_array()) {
        (*group)["lines"] = Json::array();
        lines = field(*group, "lines");
    }
    // Python keeps the matched OBJECTS alive (list identity), so stash
    // the template's payload BEFORE the replace drops the stale entries.
    Json template_copy;
    if (template_line != nullptr) template_copy = *template_line;
    Json kept = Json::array();
    for (Json& line : *lines) {
        bool stale = false;
        for (Json* m : matched) {
            if (&line == m) {
                stale = true;
                break;
            }
        }
        if (!stale) kept.push_back(line);
    }
    *lines = std::move(kept);

    const std::string layer_name = str_of(*layer, "name");
    const std::string group_horizon = str_of(*group, "target_horizon");
    const std::string kind_label =
        kind.has_value() ? constraint_kind_label(*kind) : layer_name;
    int index = 0;
    std::size_t coordinates_total = 0;
    Json all_points = Json::array();
    for (const Json& coordinates : sequences) {
        const std::string fingerprint =
            constraint_content_fingerprint(coordinates);
        Json line;
        if (index == 0 && template_line != nullptr) {
            line = template_copy;
            line["coordinates"] = coordinates;
        } else {
            line = Json::object();
            line["name"] =
                layer_name.empty() ? kind_label : layer_name;
            line["role"] =
                template_line != nullptr
                    ? str_of(template_copy, "role")
                    : std::string(
                          constraint_interpolation_role(*kind)
                              .value_or("boundary"));
            line["target_horizon"] =
                template_line != nullptr
                    ? str_of(template_copy, "target_horizon")
                    : group_horizon;
            Json props = Json::object();
            props["layer_id"] = layer_id;
            props["constraint_kind"] = *kind;
            props["feature_index"] = index;
            line["properties"] = std::move(props);
            line["coordinates"] = coordinates;
        }
        Json* props = field(line, "properties");
        if (props == nullptr || !props->is_object()) {
            line["properties"] = Json::object();
            props = field(line, "properties");
        }
        (*props)["layer_id"] = layer_id;
        (*props)["constraint_kind"] = *kind;
        (*props)["content_fingerprint"] = fingerprint;
        lines->push_back(std::move(line));
        for (const Json& point : coordinates) all_points.push_back(point);
        coordinates_total += coordinates.size();
        ++index;
    }

    report["ok"] = true;
    report["matched_by"] =
        matched_by.empty() ? "created_new_line" : matched_by;
    report["lines_synced"] = static_cast<int>(sequences.size());
    report["features_harvested"] = static_cast<int>(sequences.size());
    report["interior_rings_skipped"] = holes;
    report["coordinates_total"] = static_cast<int>(coordinates_total);
    report["content_fingerprint"] =
        constraint_content_fingerprint(all_points);
    return report;
}

}  // namespace pwb::ui_composite
